pkg/app/launcher/cmd/launcher/launcher.go

```diff
@@ -302,35 +356,109 @@ func (l *launcher) run(ctx context.Context, input cli.Input) error {
 	return nil
 }
 
+func (l *launcher) enqueueNewCommands(ctx context.Context, input cli.Input) {
+	input.Logger.Info("fetching unhandled commands to enqueue")
+
+	commands := l.commandLister.ListPipedCommands()
+	if len(commands) == 0 {
+		input.Logger.Info("there is no command to enqueue")
+		return
+	}
+
+	news := make([]model.ReportableCommand, 0, len(commands))
+	cmds := make(map[string]struct{}, len(commands))
+	for _, cmd := range commands {
+		cmds[cmd.Id] = struct{}{}
+		if _, ok := l.prevCommands[cmd.Id]; !ok {
+			news = append(news, cmd)
+		}
+	}
+
+	input.Logger.Info("fetched unhandled commands to enqueue",
+		zap.Any("pre-commands", l.prevCommands),
+		zap.Any("commands", cmds),
+		zap.Int("news", len(news)),
+	)
+
+	if len(news) == 0 {
+		input.Logger.Info("there is no new command to enqueue")
+		return
+	}
+
+	l.prevCommands = cmds
+	input.Logger.Info(fmt.Sprintf("will enqueue %d new commands", len(news)))
+
+	for _, cmd := range news {
+		select {
+		case l.commandCh <- cmd:
+			input.Logger.Info("queued a new new command", zap.String("command", cmd.Id))
+
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (l *launcher) handleCommand(ctx context.Context, input cli.Input, cmd model.ReportableCommand) {
+	logger := input.Logger.With(
+		zap.String("command", cmd.Id),
+	)
+	logger.Info("received a restart piped command to handle")
+
+	l.detectRestartCommand = true
```

Comment:
Just feel like this one is not an effective move to implement this feature, we need on ticker (to fetch the commands) and one other ticker(util launcher self checkInterval) to restart the piped. Should we rewrite some part where we restart piped by launcher to reuse that?
ref: current rerun logic https://github.com/pipe-cd/pipecd/blob/master/pkg/app/launcher/cmd/launcher/launcher.go#L248-L262

Reply:
we can keep the use of this `l.detectRestartCommand` flag but instead of forking a new go routine to just change this flag value, we can make this in the current `enqueueNewCommands`, and in execute function, we have `shouldRelaunch()` function which check for the flag and restart the piped for us 🤔